### PR TITLE
Unify all logfile handling into utils_logfile

### PR DESCRIPTION
**Pull request by pevogam**
Most importantly, for better decoupling with aexpect we provide sessions with close hooks that should have the same effect of closing the open log file descriptors but in a more decoupled fashion.

The logging functionality in utils_misc has been fully deprecated for the sake of the same functionality in a fully devoted module.

### CODE REVIEW

**Review by ldoktor — APPROVED** [review]
LGTM, seems to cover all `utils_misc.*log` occurrences while keeping the deprecated interface for `tp-*` and other consumers. As far as I understand this PR needs to be merged before https://github.com/avocado-framework/aexpect/pull/124 in order for avocado-vt to work with the aexpect after the https://github.com/avocado-framework/aexpect/pull/124 is merged, right (it will still work with the released aexpect, though)

**Comment by pevogam** [comment]
> LGTM, seems to cover all `utils_misc.*log` occurrences while keeping the deprecated interface for `tp-*` and other consumers. As far as I understand this PR needs to be merged before [avocado-framework/aexpect#124](https://github.com/avocado-framework/aexpect/pull/124) in order for avocado-vt to work with the aexpect after the [avocado-framework/aexpect#124](https://github.com/avocado-framework/aexpect/pull/124) is merged, right (it will still work with the released aexpect, though)

Exactly, this is also my understanding.

File: virttest/libvirt_vm.py
Comment (luckyh) [inline virttest/libvirt_vm.py:1476]: Okay, so to my understanding the key change of this proposal is that (logically speaking) it replaced the close hook with a VT defined one, in order to get rid of tracking the log handlers via a global variant managed by both aexpect and VT.

However I have the concern that I'm not sure if the `Spawn.close_hooks` was designed as a public attribute for endusers, since there was the `_add_close_hook` method as the setter which decorated with the leading `_`. IMO, using a private attribute/method for the replacement is just yet another workaround.

I think the critical point of this issue is about to determine either of aexpect or VT should own the log handlers by itself but not both. And in the case of we'd prefer to let VT take the ownership, the lifecycle of the log handlers would be decoupled from the session instances therefore a close hook would not always be a must to be set. On the other hand, aexpect could take over the thing as well, with the registration of the log handlers to be integrated into the `set_log_file` method.

File: virttest/utils_logfile.py
Comment (luckyh) [inline virttest/utils_logfile.py:172]: Then, if we are going to let VT manage these log handlers, I'd suggest introducing a closure to retrieve them instead of accessing with session's attribute. However, firstly we still need to confirm if the close hook could be customized for aexpect objects.

```python
def close_own_log_file(log_file):

    def hook(self):
        close_log_file(log_file)

    return hook
```

File: virttest/utils_misc.py
Comment (luckyh) [inline virttest/utils_misc.py:442]: Here missed the `filename` to be passed to the underlying.

**Review by luckyh — COMMENTED** [review]
Hi @pevogam , thank you for bringing up this proposal! it indeed improves the code quality of VT and prevents some potential issues. However I have some comments regarding the following details.

File: virttest/libvirt_vm.py
Comment (pevogam) [reply virttest/libvirt_vm.py:1476]: Let me add some quick comments here:

> However I have the concern that I'm not sure if the Spawn.close_hooks was designed as a public attribute for endusers, since there was the _add_close_hook method as the setter which decorated with the leading _. IMO, using a private attribute/method for the replacement is just yet another workaround.

Since we have a public `Spawn.close_hooks` list and not a private one like `Spawn._close_hooks` we could also assume that the hooks have always been meant to be public and customizable. The way a close hook was added privately before should not impeded that.

> I think the critical point of this issue is about to determine either of aexpect or VT should own the log handlers by itself but not both. And in the case of we'd prefer to let VT take the ownership, the lifecycle of the log handlers would be decoupled from the session instances therefore a close hook would not always be a must to be set. On the other hand, aexpect could take over the thing as well, with the registration of the log handlers to be integrated into the set_log_file method.

Indeed, there are some quite important comments on the [aexpect side](https://github.com/avocado-framework/aexpect/pull/124#discussion_r1379734217) of this where I would like to hear your opinion as well in order to conclude something more definite here. Do you agree with the idea to concentrate all logging functionality of aexpect via an output function there?

Comment (luckyh) [reply virttest/libvirt_vm.py:1476]: > Since we have a public Spawn.close_hooks list and not a private one like Spawn._close_hooks we could also assume that the hooks have always been meant to be public and customizable.

Well, as we can see there is almost not any private attribute been defined by the aexpect's classes while some of them wasn't that meaningful to be accessed publicly. Anyway, I am not that keen on the rectification of that thing so it's fine for me to accept the current approach even though it was not the one I'd advocate.

> Do you agree with the idea to concentrate all logging functionality of aexpect via an output function there?

Yeah, I'd think so since the `output_func` hook was designed in a complete and self-consistent behavior. It already gives the users the maximized flexibility to handle the line-buffered output of an execution (for whatever case of logging, monitoring, etc.), therefore I didn't see the need to replace it with something else as VT has already chosen the use.

Comment (pevogam) [reply virttest/libvirt_vm.py:1476]: > > Since we have a public Spawn.close_hooks list and not a private one like Spawn._close_hooks we could also assume that the hooks have always been meant to be public and customizable.
> 
> Well, as we can see there is almost not any private attribute been defined by the aexpect's classes while some of them wasn't that meaningful to be accessed publicly. Anyway, I am not that keen on the rectification of that thing so it's fine for me to accept the current approach even though it was not the one I'd advocate.

@ldoktor What do you think about this? At least having a public hooks list that one can add hooks to seems reasonable enough to me. @luckyh Any specific suggestions on alternative solution?

> > Do you agree with the idea to concentrate all logging functionality of aexpect via an output function there?
> 
> Yeah, I'd think so since the `output_func` hook was designed in a complete and self-consistent behavior. It already gives the users the maximized flexibility to handle the line-buffered output of an execution (for whatever case of logging, monitoring, etc.), therefore I didn't see the need to replace it with something else as VT has already chosen that use.

Ok, so I think we are all for having just the most flexible and powerful `output_func` approach to logging and any kind of output handling by aexpect. I guess if we agree on the current approach here we could look into the aexpect side about this next. So far the sibling PR ended with me mentioning about my worry about breaking changes [here](https://github.com/avocado-framework/aexpect/pull/124#discussion_r1379734217).

Comment (ldoktor) [reply virttest/libvirt_vm.py:1476]: I don't see anything wrong in adding arbitrary close hooks to `aexpect`, we do our best to execute them so as long as they are reasonable (which is on the users) it should be safe.

File: virttest/utils_misc.py
Comment (pevogam) [reply virttest/utils_misc.py:442]: Good catch!

File: virttest/utils_logfile.py
Comment (pevogam) [reply virttest/utils_logfile.py]: Indeed, this is a good extra step to get rid of the stale log file attribute of the sessions which could then later be dropped on the aexpect side in favor of the single output function. Pushed with this change in a separate commit that explains the idea and that we will have to test next.

File: virttest/libvirt_vm.py
Comment (pevogam) [reply virttest/libvirt_vm.py]: Must be resolved by https://github.com/avocado-framework/avocado-vt/pull/3782/files#r1397190844.

File: virttest/libvirt_vm.py
Comment (luckyh) [reply virttest/libvirt_vm.py]: > Any specific suggestions on alternative solution?

Ok, since now the log handlers are managed by VT completely, it would be possible to move the cleanup procedure into the same place where the sessions being closed.

https://github.com/avocado-framework/avocado-vt/blob/60ef8ea2e02bf7f824ee4428689828d1e51f6d75/virttest/env_process.py#L626

On the other side, we could consider the way about doing the cleanup via aexpect's close hooks, however, what I'd suggest there would be to have a fix or at least an enhancement for introducing something like `add_close_hook`/`on_close` in stead of manipulating the `close_hooks` list directly, which can prevent the users doing harmful things.

Comment (pevogam) [reply virttest/libvirt_vm.py]: Hi @luckyh I am not sure I fully understand the alternative suggestion here.

> Ok, since now the log handlers are managed by VT completely, it would be possible to move the cleanup procedure into the same place where the sessions being closed.

Do you have a precise diff in mind so that I can better understand what you mean?

> what I'd suggest there would be to have a fix or at least an enhancement for introducing something like add_close_hook/on_close in stead of manipulating the close_hooks list directly, which can prevent the users doing harmful things.

As this is python I doubt we can prevent the users from doing harmful things no matter what we try. Also I find the replacement of a public `session.close_hooks.append(new_hook)` with `session.add_close_hook(new_hook)` to be more like replacing one already available public function with another one unless we really have some useful logic to add to the second case.

**Comment by pevogam** [comment]
I pushed a small fix for some cases of `Failed to print_line` that were caused by the last commit incorrectly removing additional output parameters of the output function in one case. All tested now also in integration with our own products in addition to this CI.

File: virttest/libvirt_vm.py
Comment (luckyh) [reply virttest/libvirt_vm.py]: Sorry for the late reply.

> Do you have a precise diff in mind so that I can better understand what you mean?

It would be something like the following.

```patch
diff --git a/virttest/env_process.py b/virttest/env_process.py
index 156bb376..7eb43c00 100644
--- a/virttest/env_process.py
+++ b/virttest/env_process.py
@@ -44,6 +44,7 @@ from virttest import utils_qemu
 from virttest import migration
 from virttest import utils_kernel_module
 from virttest import arch
+from virttest import utils_logfile
 from virttest.utils_conn import SSHConnection
 from virttest.utils_version import VersionInterval
 from virttest.staging import service
@@ -630,6 +631,8 @@ def postprocess_vm(test, params, env, name):
         except Exception:
             pass
 
+    utils_logfile.close_all_log_files()
+
     if params.get("vm_extra_dump_paths") is not None:
         vm_extra_dumps = os.path.join(test.outputdir, "vm_extra_dumps")
         if not os.path.exists(vm_extra_dumps):
diff --git a/virttest/utils_logfile.py b/virttest/utils_logfile.py
index bd915e8a..6ae10060 100644
--- a/virttest/utils_logfile.py
+++ b/virttest/utils_logfile.py
@@ -160,3 +160,16 @@ def close_log_file(filename):
                 _open_log_files.pop(key_to_remove)
     finally:
         _log_lock.release()
+
+
+def close_all_log_files():
+    global _open_log_files, _log_file_dir, _log_lock
+    if not _acquire_lock(_log_lock):
+        raise LogLockError("Could not acquire exclusive lock to access"
+                           " _open_log_files")
+    try:
+        while _open_log_files:
+            _, log_fd = _open_log_files.popitem():
+            log_fd.close()
+    finally:
+        _log_lock.release()
```

> As this is python I doubt we can prevent the users from doing harmful things no matter what we try.

The important thing is that it can give a warning to the reviewers (no matter human or bot) for doing a check on the "defective code" when they are reading about that.

Let's just recall on how this issue was raised. The story was starting from a suspicious data access was found, right?

> Also I find the replacement of a public `session.close_hooks.append(new_hook)` with `session.add_close_hook(new_hook)` to be more like replacing one already available public function with another one unless we really have some useful logic to add to the second case.

The key point is with accessing the `close_hooks` attribute directly one could do something like `session.close_hooks.clear()` which may not be allowed to be issued by a external caller, while there is no way for the latter to do as so.

Comment (pevogam) [reply virttest/libvirt_vm.py]: > It would be something like the following.

I am not sure if this is taking this PR too far already considering its objective of just separating the dependencies but I pushed your required change with some improved reusability of already existing code.

> The key point is with accessing the `close_hooks` attribute directly one could do something like `session.close_hooks.clear()` which may not be allowed to be issued by a external caller, while there is no way for the latter to do as so.

I don't see anything wrong with a user calling `session.close_hooks.clear()` as the close hooks attribute is meant to be public and provide access to all hooks and their overall management via the list. So I don't see why an external caller should not be allowed to issue this call or how they would be mislead in issuing such a call via the API.

File: virttest/utils_logfile.py
Comment (luckyh) [inline virttest/utils_logfile.py:159]: Suggest as the following then we can get rid of adding a new function.

```python
def close_log_file(filename='*'):
    ...
        for log_file, log_fd in _open_log_files.items():
            if (filename == '*' or
                os.path.basename(log_file) == os.path.basename(filename)):
    ...
```

Comment (pevogam) [reply virttest/utils_logfile.py:159]: Do you mean `close_log_files` here as the proper name? A wildcard indicates we can close more than a single file. Also which functions you are referring to as redundant? The additional functions are all simple wrappers that either make the use of an obligatory `filename` param or don't use it all, I think all of them so far have clear name and intent.

File: virttest/libvirt_vm.py
Comment (pevogam) [reply virttest/libvirt_vm.py]: > Really? With manipulating the low-level abstraction, I mean the python built-in `list` stuff you mentioned above, users could even store and fetch pure text/binary data and then building up a brand new message bus subsystem,

This sounds more like having issues with the way python is made (which is excessively free and lacking validation I agree) and arguing for putting extra effort to fix it ourselves while in fact a lot of the language implicitly heavily relies on user, bot, and overall caller sanity.

Of course, adding extra validation and restricting usage ourselves is always welcome as long as this is easily feasible. But I am afraid right now I am out of resources for establishing a complete wrapper around the hooks list having `add_close_hook`, `remove_close_hook`, various validation of what callables are added, removed, inserted, sorted, and so on. This pull request has already taken months of reviews and further modifications I constantly have to come back to (switch context, catch up with what was done and what is the new thing that has to be done) and all I wanted was to quickly help out separating the two projects. So unless you have a specific minimal and very compact diff I could quickly adapt here that achieves and communicates what you want unfortunately I won't be able to help out much further than this.

File: virttest/utils_logfile.py
Comment (luckyh) [reply virttest/utils_logfile.py:159]: Sorry for the confusion. My intention was, we could introduce the wildcard `*` to the existing function `close_log_file` for extending its ability on closing all the log handlers managed by the `_open_log_files` list, and that could avoid adding a new function (i.e. the `close_log_files`).

The reason why I preferred as so is that, I feel like that the user may always using the `close_log_files` instead of another with having the both.

Comment (pevogam) [reply virttest/utils_logfile.py]: Alright, it is clear now and I added the changes.

**Review by luckyh — APPROVED** [review]
LGTM

**Comment by pevogam** [comment]
> Yep, looks good. I mean using `*` to match all is slightly controversial as the file can be named `*`

I think using `*` will be a fairly exotic action when it comes to Avocado VT log file naming but a good observation nevertheless. A previous push of this pull request had `None` in the usage but indeed perhaps all of this is not too important since it is only meant to be used for Avocado VT log file management to begin with.

@luckyh I have integration tested the recent changes fully (with the corresponding aexpect side adapted too) and everything checks out. As this pull request has two approvals we can either merge or ask another VT maintainer to review this pull request too if you like.

File: virttest/qemu_vm.py
Comment (PaulYuuu) [inline virttest/qemu_vm.py:3535]: Hello @pevogam, I found a problem when destroying the VM, this behavior will clean the existing log file and renew one, causing old log contents missed. Here are serial log: `tail: io-github-autotest-qemu.boot.arm64-pci/serial-serial0-avocado-vt-vm1.log: file truncated`
```
2024-01-04 23:55:42: [  OK  ] Started System Logging Service.
2024-01-04 23:55:43: [   13.918355] block dm-0: the capability attribute has been deprecated.
tail: io-github-autotest-qemu.boot.arm64-pci/serial-serial0-avocado-vt-vm1.log: file truncated
2024-01-04 23:55:47: [   18.523651] sda2: Can't mount, would change RO state
2024-01-04 23:55:52: [   22.792339] audit: type=1305 audit(1704430552.379:164): op=set audit_pid=0 old=853 auid=4294967295 ses=4294967295 subj=system_u:system_r:auditd_t:s0 res=1
```

After the VM is created, serial session also opened to record all logs. However, destroy VM with graceful_shutdown will open a new serial session to send the shutdown command, so seems at that time, `file truncated` is triggered.

And I am curious that `Tail` can close the log file itself with `def _close_log_file`, why it needs `close_hooks` here? Does it have more benefits?

Comment (pevogam) [reply virttest/qemu_vm.py:3535]: Hi @PaulYuuu, `Tail` can no longer close log files and aexpect has a unified interface that only accepts an arbitrary function for handling any outputs. All the benefits from this were discussed in this pull request as well as https://github.com/avocado-framework/aexpect/commit/2d3154582f073a9b0cdcda44d1e3dbf67d1ff703#comments and the corresponding pull request on the aexpect side.

From your description it seems to me that we are not using all the right places to set up the logging since the above changes are meant to set logging only during vm creation and the vm graceful destruction is an unwanted side effect. In particular, we want to tie the logging of a session with the lifespan of a session and in turn to the lifespan of a vm. Can you point me to the exact lines that perform the unwanted truncation of the log file? IIUC the graceful shutdown of a VM is not supposed to create a new session, is that right? Do you have the exact line in question?

Comment (ldoktor) [reply virttest/qemu_vm.py:3535]: Well, re-opening the file doesn't seem as an issue to me (eg. on double cleanup I can understand a log file reopenning), the real issue IMO sits in the `utils_logfile.py` where the file is opened as `w` where `a` would suite better (as we most certainly don't want to truncate the log files in any case.

Comment (pevogam) [reply virttest/qemu_vm.py:3535]: > Well, re-opening the file doesn't seem as an issue to me (eg. on double cleanup I can understand a log file reopenning), the real issue IMO sits in the `utils_logfile.py` where the file is opened as `w` where `a` would suite better (as we most certainly don't want to truncate the log files in any case.

I think this nailed it - all we need is the assumption of opening the log file multiple times. @PaulYuuu Do you agree with this? Are you willing to propose a PR or do you want me to do so?

Comment (PaulYuuu) [reply virttest/qemu_vm.py:3535]: Thank you @ldoktor @pevogam, I agree with you, and I am sorry for my first comment, actually `close_own_log_file` is only called once. So the reason may be that when the fd is closed, some contents are still in the cache, so the process still logs them into the file, which reopens the file.

Comment (PaulYuuu) [reply virttest/qemu_vm.py:3535]: @pevogam, I am glad to open a new one, let me try first.

Comment (PaulYuuu) [reply virttest/qemu_vm.py:3535]: Hello @ldoktor @pevogam, I sent a new PR to fix this by change mode to 'a': https://github.com/avocado-framework/avocado-vt/pull/3822, thank you all. Cc @luckyh

**Pull request merged**
